ddp-workspace/projects/ddp-sdk/src/lib/invitationCodeFormatter.directive.ts

```diff
@@ -0,0 +1,230 @@
+import { Directive, ElementRef, forwardRef, HostListener, Input } from '@angular/core';
+import { MAT_INPUT_VALUE_ACCESSOR } from '@angular/material';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+import formatErrorMsg = jasmine.formatErrorMsg;
+
+const DISPLAY_SEPARATOR = ' - ';
+const DISPLAY_INPUT_MAX_LEN = 18;
+const MAX_VALUE_LEN = 12;
+@Directive({
+    selector: 'input[invitationcode]',
+    providers: [
+        { provide: MAT_INPUT_VALUE_ACCESSOR, useExisting: InvitationCodeFormatterDirective },
+        {
+            provide: NG_VALUE_ACCESSOR,
+            useExisting: forwardRef(() => InvitationCodeFormatterDirective),
+            multi: true,
+        }
+    ]
+})
+
+export class InvitationCodeFormatterDirective implements ControlValueAccessor {
+    private _value: string | null = '';
```

Comment:
This is the directive that is not quite working right yet, so don't waste too much time looking at it.
Supposed to take in the entered string and format it by uppercasing and introducing separators every 4 characters.
That part is easy.
The part that is harder is to deal updating the insertion point in the text box after someone has typed or pasted text from anywhere in the input box.
Again, have not resolved all the issues there, but at least I understand them.